Could not install reboundx due to 'wheels error'

I tried installing reboundx in the Spyder (5.4.3) terminal with python 3.11 but got a "could not build wheels for reboundx" error.

I am running macOS (14.4) and I have REBOUND 4.4.1 and setuptools (70.1.1) installed.

Here is the terminal output:


```
In [3]: pip install reboundx

Collecting reboundx

  Using cached reboundx-4.2.0.tar.gz (92 kB)

  Installing build dependencies ... [?25ldone

etting requirements to build wheel ... e

g metadata (pyproject.toml) ... [?25ldone

irement already satisfied: rebound>=4.0.0 in /opt/homebrew/anaconda3/lib/python3.11/site-packages (from reboundx) (4.4.1)

Building wheels for collected packages: reboundx

  Building wheel for reboundx (pyproject.toml) ... [?25lerror

  error: subprocess-exited-with-error


  × Building wheel for reboundx (pyproject.toml) did not run successfully.

  │ exit code: 1

  ╰─> [60 lines of output]

      fatal: not a git repository (or any of the parent directories): .git

      running bdist_wheel

      running build

      running build_py

      creating build

      creating build/lib.macosx-11.1-arm64-cpython-311

      creating build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/extras.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/params.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/tools.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/constants.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/__init__.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/simulationarchive.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/testing.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      copying reboundx/data.py -> build/lib.macosx-11.1-arm64-cpython-311/reboundx

      running build_ext

      *** /private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound *** /opt/homebrew/anaconda3/lib/python3.11/site-packages/ ***

      building 'libreboundx' extension

      creating build/temp.macosx-11.1-arm64-cpython-311

      creating build/temp.macosx-11.1-arm64-cpython-311/src

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/central_force.c -o build/temp.macosx-11.1-arm64-cpython-311/src/central_force.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/core.c -o build/temp.macosx-11.1-arm64-cpython-311/src/core.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      src/core.c:557:12: warning: 9 enumeration values not handled in switch: 'REB_INTEGRATOR_SEI', 'REB_INTEGRATOR_LEAPFROG', 'REB_INTEGRATOR_NONE'... [-Wswitch]

          switch(sim->integrator){

                 ^~~~~~~~~~~~~~~

      src/core.c:557:12: note: add missing switch cases

          switch(sim->integrator){

                 ^

      src/core.c:857:18: warning: incompatible pointer types passing 'void **' to parameter of type 'struct rebx_node **' [-Wincompatible-pointer-types]

          rebx_free_ap(&p->ap);

                       ^~~~~~

      src/core.c:845:38: note: passing argument to parameter 'ap' here

      void rebx_free_ap(struct rebx_node** ap){

                                           ^

      2 warnings generated.

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/exponential_migration.c -o build/temp.macosx-11.1-arm64-cpython-311/src/exponential_migration.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      src/exponential_migration.c:79:18: warning: unused variable 'r2' [-Wunused-variable]

          const double r2 = dx*dx + dy*dy + dz*dz;

                       ^

      1 warning generated.

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/gas_dynamical_friction.c -o build/temp.macosx-11.1-arm64-cpython-311/src/gas_dynamical_friction.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      src/gas_dynamical_friction.c:126:28: warning: using integer absolute value function 'abs' when argument is of floating point type [-Wabsolute-value]

              const double vert=(abs(diff.z)<(10*h))?exp(-diff.z*diff.z/(2.0*h*h)):0;

                                 ^

      src/gas_dynamical_friction.c:126:28: note: use function 'fabs' instead

              const double vert=(abs(diff.z)<(10*h))?exp(-diff.z*diff.z/(2.0*h*h)):0;

                                 ^~~

                                 fabs

      1 warning generated.

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/gr.c -o build/temp.macosx-11.1-arm64-cpython-311/src/gr.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      clang -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -fPIC -O2 -isystem /opt/homebrew/anaconda3/include -arch arm64 -DLIBREBOUNDX -Isrc -I/opt/homebrew/anaconda3/include/python3.11 -I/private/var/folders/yl/0_w56qsd0135bwfg4q_4v8km0000gn/T/pip-build-env-11ory4bm/overlay/lib/python3.11/site-packages/rebound -c src/gr_full.c -o build/temp.macosx-11.1-arm64-cpython-311/src/gr_full.o -fstrict-aliasing -O3 -std=c99 -Wno-unknown-pragmas -DREBXGITHASH=4b1377fe3337271cdd67013199ae2fd792ed5601 -DLIBREBOUNDX -D_GNU_SOURCE -fPIC

      src/gr_full.c:278:27: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]

                          const dotproduct = dxij*ps_b[j].ax+dyij*ps_b[j].ay+dzij*ps_b[j].az;

                          ~~~~~ ^

                          int

      src/gr_full.c:374:20: warning: variable 'sumk' set but not used [-Wunused-but-set-variable]

                  double sumk = 0.;

                         ^

      1 warning and 1 error generated.

      error: command '/usr/bin/clang' failed with exit code 1

      [end of output]


  note: This error originates from a subprocess, and is likely not a problem with pip.

  ERROR: Failed building wheel for reboundx

ailed to build reboundx

ERROR: Could not build wheels for reboundx, which is required to install pyproject.toml-based projects

Note: you may need to restart the kernel to use updated packages.



In [4]: 
```

--- a/src/core.c
+++ b/src/core.c
@@ -575,7 +575,10 @@ int rebx_add_operator(struct rebx_extras* rebx, struct rebx_operator* operator){
                 reb_simulation_error(sim, "REBOUNDx Error: Operators that affect particle trajectories are not supported with Mercurius. Must add as forces.\n");
                 return 0;
             }
+            break;
         }
+        default:
+            break;
     }
     return 0; // didn't reach a successful outcome
 }
@@ -854,7 +857,7 @@ void rebx_free_ap(struct rebx_node** ap){
 }
 
 void rebx_free_particle_ap(struct reb_particle* p){
-    rebx_free_ap(&p->ap);
+    rebx_free_ap((struct rebx_node **)(&p->ap));
 }
 
 void rebx_free_force(struct rebx_extras* rebx, struct rebx_force* force){

--- a/src/exponential_migration.c
+++ b/src/exponential_migration.c
@@ -73,10 +73,10 @@ static struct  reb_vec3d rebx_calculate_modify_orbits_forces_new(struct reb_simu
     const double dvx = p->vx - source->vx;
     const double dvy = p->vy - source->vy;
     const double dvz = p->vz - source->vz;
-    const double dx = p->x-source->x;
-    const double dy = p->y-source->y;
-    const double dz = p->z-source->z;
-    const double r2 = dx*dx + dy*dy + dz*dz;
+    //const double dx = p->x-source->x;
+    //const double dy = p->y-source->y;
+    //const double dz = p->z-source->z;
+    //const double r2 = dx*dx + dy*dy + dz*dz;
     
     if(em_tau_a_ptr != NULL){
         em_tau_a = *em_tau_a_ptr;

--- a/src/gas_dynamical_friction.c
+++ b/src/gas_dynamical_friction.c
@@ -123,7 +123,7 @@ static void rebx_calculate_gas_dynamical_friction(struct reb_simulation* const s
         //scale height is defined by user-defined aspect ratio. Truncate the disc vertically
         //at 10 scale heights...
         const double h=hr*rcyl;
-        const double vert=(abs(diff.z)<(10*h))?exp(-diff.z*diff.z/(2.0*h*h)):0;
+        const double vert=(fabs(diff.z)<(10*h))?exp(-diff.z*diff.z/(2.0*h*h)):0;
         const double rhog_loc=rhog*pow(rcyl, alpha_rhog)*vert;
         const double mp = p.m;
         const double rstar = p.r;

--- a/src/gr_full.c
+++ b/src/gr_full.c
@@ -58,80 +58,82 @@
 #include "rebound.h"
 #include "reboundx.h"
 
-static void reb_particles_transform_inertial_to_barycentric_posvelacc(const struct reb_particle* const particles, struct reb_particle* const p_b, const unsigned int N, const unsigned int N_active){
-    double x0  = 0.;
-    double y0  = 0.;
-    double z0  = 0.;
-    double vx0 = 0.;
-    double vy0 = 0.;
-    double vz0 = 0.;
-    double ax0 = 0.;
-    double ay0 = 0.;
-    double az0 = 0.;
-    double m0  = 0.;
-#pragma omp parallel for reduction(+:x0) reduction(+:y0) reduction(+:z0) reduction(+:vx0) reduction(+:vy0) reduction(+:vz0) reduction(+:ax0) reduction(+:ay0) reduction(+:az0) reduction(+:m0)
-    for (unsigned int i=0;i<N_active;i++){
-        double m = particles[i].m;
-        x0  += particles[i].x *m;
-        y0  += particles[i].y *m;
-        z0  += particles[i].z *m;
-        vx0 += particles[i].vx*m;
-        vy0 += particles[i].vy*m;
-        vz0 += particles[i].vz*m;
-        ax0 += particles[i].ax*m;
-        ay0 += particles[i].ay*m;
-        az0 += particles[i].az*m;
-        m0  += m;
-    }
-    p_b[0].x  = x0/m0;
-    p_b[0].y  = y0/m0;
-    p_b[0].z  = z0/m0;
-    p_b[0].vx = vx0/m0;
-    p_b[0].vy = vy0/m0;
-    p_b[0].vz = vz0/m0;
-    p_b[0].ax = ax0/m0;
-    p_b[0].ay = ay0/m0;
-    p_b[0].az = az0/m0;
-    p_b[0].m = m0;
-    
-#pragma omp parallel for 
-    for (unsigned int i=1;i<N;i++){
-        p_b[i].x  = particles[i].x  - p_b[0].x;
-        p_b[i].y  = particles[i].y  - p_b[0].y;
-        p_b[i].z  = particles[i].z  - p_b[0].z;
-        p_b[i].vx = particles[i].vx - p_b[0].vx;
-        p_b[i].vy = particles[i].vy - p_b[0].vy;
-        p_b[i].vz = particles[i].vz - p_b[0].vz;
-        p_b[i].ax = particles[i].ax - p_b[0].ax;
-        p_b[i].ay = particles[i].ay - p_b[0].ay;
-        p_b[i].az = particles[i].az - p_b[0].az;
-        p_b[i].m  = particles[i].m;
-    }
-}
-
-static void reb_particles_transform_barycentric_to_inertial_acc(struct reb_particle* const particles, const struct reb_particle* const p_b, const unsigned int N, const unsigned int N_active){
-    const double mtot = p_b[0].m;
-    double ax0  = 0.;
-    double ay0  = 0.;
-    double az0  = 0.;
-#pragma omp parallel for reduction(+:ax0) reduction(+:ay0) reduction(+:az0) 
-    for (unsigned int i=1;i<N_active;i++){
-        double m = p_b[i].m;
-        ax0 += p_b[i].ax*m/mtot;
-        ay0 += p_b[i].ay*m/mtot;
-        az0 += p_b[i].az*m/mtot;
-        particles[i].m = m; // in case of merger/mass change
-    }
-    particles[0].ax  = p_b[0].ax - ax0;
-    particles[0].ay  = p_b[0].ay - ay0;
-    particles[0].az  = p_b[0].az - az0;
-#pragma omp parallel for 
-    for (unsigned int i=1;i<N;i++){
-        particles[i].ax = p_b[i].ax+particles[0].ax;
-        particles[i].ay = p_b[i].ay+particles[0].ay;
-        particles[i].az = p_b[i].az+particles[0].az;
-    }
-}
+// The following functions are not used.
+//
+//static void reb_particles_transform_inertial_to_barycentric_posvelacc(const struct reb_particle* const particles, struct reb_particle* const p_b, const unsigned int N, const unsigned int N_active){
+//    double x0  = 0.;
+//    double y0  = 0.;
+//    double z0  = 0.;
+//    double vx0 = 0.;
+//    double vy0 = 0.;
+//    double vz0 = 0.;
+//    double ax0 = 0.;
+//    double ay0 = 0.;
+//    double az0 = 0.;
+//    double m0  = 0.;
+//#pragma omp parallel for reduction(+:x0) reduction(+:y0) reduction(+:z0) reduction(+:vx0) reduction(+:vy0) reduction(+:vz0) reduction(+:ax0) reduction(+:ay0) reduction(+:az0) reduction(+:m0)
+//    for (unsigned int i=0;i<N_active;i++){
+//        double m = particles[i].m;
+//        x0  += particles[i].x *m;
+//        y0  += particles[i].y *m;
+//        z0  += particles[i].z *m;
+//        vx0 += particles[i].vx*m;
+//        vy0 += particles[i].vy*m;
+//        vz0 += particles[i].vz*m;
+//        ax0 += particles[i].ax*m;
+//        ay0 += particles[i].ay*m;
+//        az0 += particles[i].az*m;
+//        m0  += m;
+//    }
+//    p_b[0].x  = x0/m0;
+//    p_b[0].y  = y0/m0;
+//    p_b[0].z  = z0/m0;
+//    p_b[0].vx = vx0/m0;
+//    p_b[0].vy = vy0/m0;
+//    p_b[0].vz = vz0/m0;
+//    p_b[0].ax = ax0/m0;
+//    p_b[0].ay = ay0/m0;
+//    p_b[0].az = az0/m0;
+//    p_b[0].m = m0;
+//    
+//#pragma omp parallel for 
+//    for (unsigned int i=1;i<N;i++){
+//        p_b[i].x  = particles[i].x  - p_b[0].x;
+//        p_b[i].y  = particles[i].y  - p_b[0].y;
+//        p_b[i].z  = particles[i].z  - p_b[0].z;
+//        p_b[i].vx = particles[i].vx - p_b[0].vx;
+//        p_b[i].vy = particles[i].vy - p_b[0].vy;
+//        p_b[i].vz = particles[i].vz - p_b[0].vz;
+//        p_b[i].ax = particles[i].ax - p_b[0].ax;
+//        p_b[i].ay = particles[i].ay - p_b[0].ay;
+//        p_b[i].az = particles[i].az - p_b[0].az;
+//        p_b[i].m  = particles[i].m;
+//    }
+//}
+//
+//static void reb_particles_transform_barycentric_to_inertial_acc(struct reb_particle* const particles, const struct reb_particle* const p_b, const unsigned int N, const unsigned int N_active){
+//    const double mtot = p_b[0].m;
+//    double ax0  = 0.;
+//    double ay0  = 0.;
+//    double az0  = 0.;
+//#pragma omp parallel for reduction(+:ax0) reduction(+:ay0) reduction(+:az0) 
+//    for (unsigned int i=1;i<N_active;i++){
+//        double m = p_b[i].m;
+//        ax0 += p_b[i].ax*m/mtot;
+//        ay0 += p_b[i].ay*m/mtot;
+//        az0 += p_b[i].az*m/mtot;
+//        particles[i].m = m; // in case of merger/mass change
+//    }
+//    particles[0].ax  = p_b[0].ax - ax0;
+//    particles[0].ay  = p_b[0].ay - ay0;
+//    particles[0].az  = p_b[0].az - az0;
+//#pragma omp parallel for 
+//    for (unsigned int i=1;i<N;i++){
+//        particles[i].ax = p_b[i].ax+particles[0].ax;
+//        particles[i].ay = p_b[i].ay+particles[0].ay;
+//        particles[i].az = p_b[i].az+particles[0].az;
+//    }
+//}
 
 static void rebx_calculate_gr_full(struct reb_simulation* const sim, struct reb_particle* const particles, const int N, const double C2, const double G, const int max_iterations, const int gravity_ignore_10){
     
@@ -275,7 +277,7 @@ static void rebx_calculate_gr_full(struct reb_simulation* const sim, struct reb_
                     const double dzij = ps_b[i].z - ps_b[j].z;
                     const double rij = sqrt(dxij*dxij + dyij*dyij + dzij*dzij);
                     const double rij3 = rij*rij*rij;
-                    const dotproduct = dxij*ps_b[j].ax+dyij*ps_b[j].ay+dzij*ps_b[j].az;
+                    const double dotproduct = dxij*ps_b[j].ax+dyij*ps_b[j].ay+dzij*ps_b[j].az;
 
                     non_constx += (G*particles[j].m*dxij/rij3)*dotproduct/(2.*C2) + (7./(2.*C2))*G*particles[j].m*ps_b[j].ax/rij;
                     non_consty += (G*particles[j].m*dyij/rij3)*dotproduct/(2.*C2) + (7./(2.*C2))*G*particles[j].m*ps_b[j].ay/rij;
@@ -370,18 +372,19 @@ double rebx_gr_full_hamiltonian(struct rebx_extras* const rebx, const struct reb
             
             double vtildei2 = vtilde[i].x*vtilde[i].x + vtilde[i].y*vtilde[i].y + vtilde[i].z*vtilde[i].z;
             double A = (1. - 0.5*vtildei2/C2);
-            
-            double sumk = 0.;
-            for (int k=0;k<N;k++){
-                if (k!=i){
-                    struct reb_particle pk = particles[k];
-                    double xik = pk.x - pi.x;
-                    double yik = pk.y - pi.y;
-                    double zik = pk.z - pi.z;
-                    double rik = sqrt(xik*xik + yik*yik + zik*zik);
-                    sumk -= 2.*G*pk.m/rik;
-                }
-            }
+           
+            // The result of the following calculation is never used.
+            //double sumk = 0.;
+            //for (int k=0;k<N;k++){
+            //    if (k!=i){
+            //        struct reb_particle pk = particles[k];
+            //        double xik = pk.x - pi.x;
+            //        double yik = pk.y - pi.y;
+            //        double zik = pk.z - pi.z;
+            //        double rik = sqrt(xik*xik + yik*yik + zik*zik);
+            //        sumk -= 2.*G*pk.m/rik;
+            //    }
+            //}
             
             struct reb_vec3d dv_pn = {0.};
             for (int j=0;j<N;j++){

--- a/src/input.c
+++ b/src/input.c
@@ -376,7 +376,7 @@ static int rebx_load_particle(struct rebx_extras* rebx, FILE* inf, enum rebx_inp
         switch (field.type){
             case REBX_BINARY_FIELD_TYPE_PARAM_LIST:
             {
-                if (!rebx_load_list(rebx, REBX_BINARY_FIELD_TYPE_PARAM, &p->ap, inf, warnings)){
+                if (!rebx_load_list(rebx, REBX_BINARY_FIELD_TYPE_PARAM, (struct rebx_node **)(&p->ap), inf, warnings)){
                     return 0;
                 }
                 break;

--- a/src/output.c
+++ b/src/output.c
@@ -252,6 +252,9 @@ static void rebx_write_list(struct rebx_extras* rebx, enum rebx_binary_field_typ
                 rebx_write_step(rebx, current->object, of);
                 break;
             }
+            default:
+                // Should implement a check for other types here
+                break;
         }
         N--;
     }

